### ISSUE TITLE
Inconsistent behaviour with attribute-defined-outside-init

### ISSUE THREAD

**Issue by gstorer** [PyCQA/pylint#5214]
### Bug description
There seems to be an inconsistency in the way pylint treats attributes that have been defined in `__init__` and a helper function called by `__init__`.

I would expect the two assignments to var1 would either both trigger the linter or both be OK. Instead the first is OK and the second isn't.

```python

"""Module docstring"""

class TestLint:
    """Test doc string"""

    def __init__(self):
        self.init_helper()
        self.var2 = True

    def init_helper(self):
        """Helper doc string"""
        self.var1 = True # no lint error

    def other_func(self):
        """Other func doc string"""
        self.var1 = False # lint error
        self.var2 = False # no lint error
```

### Command used

```shell
pylint test.py
```

### Pylint output

```shell
test.py:16:8: W0201: Attribute 'var1' defined outside __init__ (attribute-defined-outside-init)
```

### Expected behavior

I am not sure which of the two options is correct but the current behavior with differing behavior is wrong.

### Pylint version

```shell
pylint 2.11.1
astroid 2.8.4
Python 3.8.10 (default, Sep 28 2021, 16:10:42) 
[GCC 9.3.0]
```

**Comment by edrosten** [comment]
@DanielNoord I've opened a PR.


> Ideally `pylint` would recognise that `init_helper()` gets unconditionally called by `__init__` thus `self.var1` will always be defined before `other_func` gets called.

I've not implemented this. The reason is for consistency with existing pylint. Consider:

```python
class A():
  def __init__(self):
    if random.random() > 0.5:
      self.a=True
  
  def func(self):
      self.a=False # No lint error.
```

I figured since we don't do flow control on variable assignment, it would be inconsistent to add it on helper functions too, so that means we don't really need control flow here. 

Note that I've also not implemented helpers-of-helpers either. I don't think that's especially difficult, but pylint doesn't currently implement that either.


@Pierre-Sassoulas I don't think this needs control flow label, at least not this time round.
